domains/emissions/Emissions.API/API/Services/SourcesCalculator.cs

```diff
@@ -0,0 +1,152 @@
+﻿using API.Helpers;
+using API.Models;
+
+namespace API.Services
+{
+    public class SourcesCalculator : ISourcesCalculator
+    {
+        readonly IList<string> renewableSources = Configuration.GetRenewableSources();
+        readonly float wasteRenewableShare = Configuration.GetWasteRenewableShare();
+
+        public EnergySourceResponse CalculateSourceEmissions(
+            IEnumerable<TimeSeries> timeSeries,
+            List<Record> declarationRecords,
+            Aggregation aggregation)
+        {
+            var groupedDeclarations = GetGroupedDeclarations(aggregation, declarationRecords);
+            var consumptionResults = new Dictionary<string, Dictionary<string, ConsumptionShare>>();
+
+            foreach (var measurements in timeSeries)
+            {
+                foreach (var measurement in measurements.Measurements)
+                {
+                    var utcDateTime = GetDateAsString(measurement.DateFrom.ToUtcDateTime(), aggregation);
+                    var gridArea = measurements.MeteringPoint.GridArea;
+                    var totalShares = groupedDeclarations[utcDateTime + gridArea];
+
+                    CalculateConsumptionShare(consumptionResults, utcDateTime, totalShares, measurement);
+                }
+            }
+            var result = CalculateSourceEmissionPercentage(timeSeries, aggregation, consumptionResults);
+
+            return result;
+        }
+
+        void CalculateConsumptionShare(
+            Dictionary<string, Dictionary<string, ConsumptionShare>> consumptionResults,
+            string utcDateTime,
+            IEnumerable<Record> totalShares,
+            Measurement measurement)
+        {

```

Comment:
Let us rework this method to return its result instead of manipulating an existing object. Alternatively, *if* that is very hard to do, rename the method to clearly indicate that it is updating the `results` object.

Reply:
An issuse will be made to rewrite the file.